src/main/kotlin/com/exactpro/th2/httpserver/server/Th2HttpServer.kt

```diff
@@ -13,35 +13,121 @@
+                    executorService.submit { handle(socket.accept()) }
+                } catch (e: SocketException) {
+                    LOGGER.error(e) { "Broken socket, trying to recreate..." }
+                    recreateSocket()
+                } catch (e: IOException) {
+                    LOGGER.error(e) { "Failed accept" }
+                    if (socket.isClosed) recreateSocket()
+                }
+            }
+        }, "th2-conn-http-server").start()
+    }
+
+    private fun recreateSocket() {
+        options.runCatching { socket = createSocket() }.onFailure { e -> LOGGER.error(e) { "Can't recreate socket!" }}
+    }
+
+    private fun handle(client: Socket) {
+        var request: RawHttpRequest
+        var serverWillCloseConnection = false
+        while (!serverWillCloseConnection) {
+            try {
+                request = http.parseRequest(
+                    client.getInputStream(),
+                    (client.remoteSocketAddress as InetSocketAddress).address
+                )
+
+                val requestEagerly = request.eagerly()
+                options.onRequest(requestEagerly)
+
+                val connectionOption = request.headers.getFirst("Connection")
+
+                serverWillCloseConnection = connectionOption.map { string: String? -> "close".equals(string, ignoreCase = true) }.orElse(false)
+
+                if (!serverWillCloseConnection) serverWillCloseConnection = !keepAlive(request.startLine.httpVersion, connectionOption)
+
+                LOGGER.debug("Request: \n$requestEagerly\nwas send from client$")
+
+                onRequest(requestEagerly) { res: RawHttpResponse<*> ->
+                    val response = options.prepareResponse(requestEagerly, res).apply { writeTo(client.getOutputStream()) }
+                    LOGGER.debug("Response: \n$response\nwas send to client")
+                    options.onResponse(requestEagerly, response)
+                    closeBodyOf(response)
+                }
+            } catch (e: Exception) {
+                if (e !is SocketException) {
+                    // only print stack trace if this is not due to a client closing the connection
+                    val clientClosedConnection = e is InvalidHttpRequest && e.lineNumber == 0
+                    if (!clientClosedConnection) LOGGER.error(e.stackTraceToString())
+                    client.runCatching(Socket::close)
+                }
+                serverWillCloseConnection = true // cannot keep listening anymore
+            } finally {
+                if (serverWillCloseConnection) client.runCatching(Socket::close)
+            }
         }
     }
 
     override fun stop() {
-        val router = routerRef.getAndSet(null)
-        router?.run {
-            stop()
-            logger.info("Server is stopped")
+        try {
+            socket.close()
+        } catch (e: IOException) {
+            LOGGER.warn(e) {}
+        } finally {
+            executorService.shutdown()
```

Comment:
Usually executor shutdown sequence goes like this:
1. call `shutdown`
2. call `await` with some timeout
3. if `await` returned `false` call `shutdownNow` and log a message that executor wasn't shutdown in a specified timeout